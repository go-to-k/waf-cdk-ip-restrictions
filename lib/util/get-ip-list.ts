import * as fs from "fs";

const ipListFilePath = "./iplist.txt";

export const getIPList = (): string[] => {
  const ipList: string[] = [];

  const ipListFile = fs.readFileSync(ipListFilePath, "utf8");
  const lines = ipListFile.toString().split("\n");

  for (const line of lines) {
    // TODO: check CIDR format or not (/32, /16, etc...)
    const trimmedLine = line
      .replace(/ /g, "")
      .replace(/\t/g, "")
      .replace(/\n/g, "")
      .replace(/^([^#]+)#.*$/g, "$1");

    const pattern = /^#/g;
    const result = trimmedLine.match(pattern);

    if (trimmedLine.length && !result) {
      ipList.push(trimmedLine);
    }
  }

  return ipList;
};
