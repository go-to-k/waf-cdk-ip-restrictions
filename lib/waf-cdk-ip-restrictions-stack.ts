import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import * as wafv2 from "aws-cdk-lib/aws-wafv2";
import * as fs from "fs";

const scopeType = process.env.SCOPE_TYPE ?? "";
const ipListFilePath = "./iplist.txt";
export class WafCdkIpRestrictionsStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.validate(scopeType, props?.env?.region ?? "");

    const ipList = this.getIPList();

    const whiteListIPSet = new wafv2.CfnIPSet(this, "WhiteListIPSet", {
      name: "WhiteListIPSet",
      addresses: ipList,
      ipAddressVersion: "IPV4",
      scope: scopeType,
    });

    const whiteListIPSetRuleProperty: wafv2.CfnWebACL.RuleProperty = {
      priority: 0,
      name: "WhiteListIPSet-Rule",
      action: {
        allow: {},
      },
      statement: {
        ipSetReferenceStatement: {
          arn: whiteListIPSet.attrArn,
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: "WhiteListIPSet-Rule",
        sampledRequestsEnabled: true,
      },
    };

    const webAcl = new wafv2.CfnWebACL(this, "WebAcl", {
      name: "WebAcl",
      defaultAction: { block: {} },
      scope: scopeType,
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: "WebAcl",
        sampledRequestsEnabled: true,
      },
      rules: [whiteListIPSetRuleProperty],
    });
  }

  private validate(scopeType: string, region: string): void {
    if (scopeType !== "CLOUDFRONT" && scopeType !== "REGIONAL") {
      throw new Error("Scope must be CLOUDFRONT or REGIONAL.");
    }
    if (scopeType === "CLOUDFRONT" && region !== "us-east-1") {
      throw new Error("Region must be us-east-1 when CLOUDFRONT.");
    }
  }

  // TODO: move this to utils dir, and make unit tests (for public method)
  private getIPList(): string[] {
    const ipList: string[] = [];

    const ipListFile = fs.readFileSync(ipListFilePath, "utf8");
    const lines = ipListFile.toString().split("\r\n");

    for (const line of lines) {
      const trimmedLine = line
        .replace(/ /g, "")
        .replace(/\t/g, "")
        .replace(/\n/g, "")
        .replace(/^([^#]+)#.*$/g, "$1");

      // TODO: check empty lines whether or not (need \n??)
      const pattern = /^#/g;
      const result = trimmedLine.match(pattern);

      if (trimmedLine.length && !result) {
        ipList.push(line);
      }
    }

    return ipList;
  }
}
