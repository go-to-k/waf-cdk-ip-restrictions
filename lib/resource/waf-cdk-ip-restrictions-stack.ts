import { Stack } from "aws-cdk-lib";
import { Construct } from "constructs";
import { getIPList } from "../util/get-ip-list";
import { WafRegionValidator } from "../validator/waf-region-validator";
import { CfnIPSet, CfnWebACL } from "aws-cdk-lib/aws-wafv2";
import { ConfigStackProps } from "../config";

const ipListFilePath = "./iplist.txt";

export class WafCdkIpRestrictionsStack extends Stack {
  private scopeType: string;
  private ipList: string[];

  constructor(scope: Construct, id: string, props: ConfigStackProps) {
    super(scope, id, props);

    this.init(props);
    this.create();
  }

  private init(props: ConfigStackProps): void {
    this.scopeType = props.config.scopeType;
    this.ipList = getIPList(ipListFilePath);

    const wafRegionValidator = new WafRegionValidator(this.scopeType, this.region);
    this.node.addValidation(wafRegionValidator);
  }

  private create(): void {
    const whiteListIPSet = new CfnIPSet(this, "WhiteListIPSet", {
      name: "WhiteListIPSet",
      addresses: this.ipList,
      ipAddressVersion: "IPV4",
      scope: this.scopeType,
    });

    const whiteListIPSetRuleProperty: CfnWebACL.RuleProperty = {
      priority: 0,
      name: "WhiteListIPSet-Rule",
      action: {
        allow: {},
      },
      statement: {
        ipSetReferenceStatement: {
          arn: whiteListIPSet.attrArn,
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: "WhiteListIPSet-Rule",
        sampledRequestsEnabled: true,
      },
    };

    new CfnWebACL(this, "WebAcl", {
      name: "WebAcl",
      defaultAction: { block: {} },
      scope: this.scopeType,
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: "WebAcl",
        sampledRequestsEnabled: true,
      },
      rules: [whiteListIPSetRuleProperty],
    });
  }
}
