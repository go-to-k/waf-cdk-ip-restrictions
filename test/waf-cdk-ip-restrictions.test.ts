import * as cdk from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import { WafCdkIpRestrictionsStack } from "../lib/resource/waf-cdk-ip-restrictions-stack";
import _cdkJsonRaw from "../cdk.json";
import { getIPList } from "../lib/util/get-ip-list";

export type CdkJson = typeof _cdkJsonRaw;

const getTemplate = (region?: string, scopeType?: string): cdk.assertions.Template => {
  const _cdkJson: CdkJson = _cdkJsonRaw;

  if (region !== undefined) _cdkJson.context.region = region;
  if (scopeType !== undefined) _cdkJson.context.scopeType = scopeType;

  const app = new cdk.App({
    context: _cdkJson.context,
  });
  const regionContext = app.node.tryGetContext("region") ?? "";
  const stack = new WafCdkIpRestrictionsStack(app, "WafCdkIpRestrictionsStack", {
    env: {
      region: regionContext,
    },
  });
  const template = Template.fromStack(stack);
  return template;
};

describe("Snapshot Tests", () => {
  const template = getTemplate();

  test("Snapshot test", () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});

describe("Fine-grained Assertions Tests", () => {
  const template = getTemplate();

  test("WebACL created", () => {
    template.resourceCountIs("AWS::WAFv2::WebACL", 1);
  });

  test("IPSet created", () => {
    template.resourceCountIs("AWS::WAFv2::IPSet", 1);
  });

  test("WAF default action is block", () => {
    template.hasResourceProperties("AWS::WAFv2::WebACL", {
      DefaultAction: { Block: {} },
    });
  });

  test("IPSet has addresses", () => {
    template.hasResourceProperties("AWS::WAFv2::IPSet", {
      Addresses: Match.anyValue(),
    });
  });
});

describe("Validation Tests", () => {
  test("SCOPE: CLOUDFRONT - REGION: ap-northeast-1 = Error", () => {
    const region = "ap-northeast-1";
    const scopeType = "CLOUDFRONT";
    expect(() => {
      const template = getTemplate(region, scopeType);
    }).toThrowError(/Region must be us-east-1 when CLOUDFRONT./);
  });

  test("SCOPE: CLOUDFRONT - REGION: us-east-1 = Success", () => {
    const region = "us-east-1";
    const scopeType = "CLOUDFRONT";
    expect(() => {
      const template = getTemplate(region, scopeType);
    }).not.toThrow(Error);
  });

  test("SCOPE: REGIONAL - REGION: ap-northeast-1 = Success", () => {
    const region = "ap-northeast-1";
    const scopeType = "REGIONAL";
    expect(() => {
      const template = getTemplate(region, scopeType);
    }).not.toThrow(Error);
  });

  test("SCOPE: REGIONAL - REGION: us-east-1 = Success", () => {
    const region = "us-east-1";
    const scopeType = "REGIONAL";
    expect(() => {
      const template = getTemplate(region, scopeType);
    }).not.toThrow(Error);
  });

  test("SCOPE: INVALID - REGION: us-east-1 = Error", () => {
    const region = "us-east-1";
    const scopeType = "INVALID";
    expect(() => {
      const template = getTemplate(region, scopeType);
    }).toThrowError(/Scope must be CLOUDFRONT or REGIONAL./);
  });
});

describe("IP List Tests", () => {
  test("iplist-1 = Success", () => {
    const ipListFilePath = "./test/iplists/iplist-1.txt";
    expect(getIPList(ipListFilePath)).toEqual([]);
  });

  test("iplist-2 = Error", () => {
    const ipListFilePath = "./test/iplists/iplist-2.txt";
    expect(() => {
      getIPList(ipListFilePath);
    }).toThrowError(/IP CIDR Format is invalid:/);
  });

  test("iplist-3 = Error", () => {
    const ipListFilePath = "./test/iplists/iplist-3.txt";
    expect(() => {
      getIPList(ipListFilePath);
    }).toThrowError(/IP CIDR Format is invalid:/);
  });

  test("iplist-4 = Success", () => {
    const ipListFilePath = "./test/iplists/iplist-4.txt";
    expect(getIPList(ipListFilePath)).toEqual(["0.0.0.1/16", "0.0.0.2/32", "0.0.0.3/24"]);
  });

  test("iplist-5 = Error", () => {
    const ipListFilePath = "./test/iplists/iplist-5.txt";
    expect(() => {
      getIPList(ipListFilePath);
    }).toThrowError(/IP CIDR Format is invalid:/);
  });

  test("iplist-6 = Error", () => {
    const ipListFilePath = "./test/iplists/iplist-6.txt";
    expect(() => {
      getIPList(ipListFilePath);
    }).toThrowError(/IP CIDR Format is invalid:/);
  });
});
