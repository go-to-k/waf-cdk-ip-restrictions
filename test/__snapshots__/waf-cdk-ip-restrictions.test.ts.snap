// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests Snapshot test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "WebAcl": Object {
      "Properties": Object {
        "DefaultAction": Object {
          "Block": Object {},
        },
        "Name": "WebAcl",
        "Rules": Array [
          Object {
            "Action": Object {
              "Allow": Object {},
            },
            "Name": "WhiteListIPSet-Rule",
            "Priority": 0,
            "Statement": Object {
              "IPSetReferenceStatement": Object {
                "Arn": Object {
                  "Fn::GetAtt": Array [
                    "WhiteListIPSet",
                    "Arn",
                  ],
                },
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "WhiteListIPSet-Rule",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUDFRONT",
        "VisibilityConfig": Object {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "WebAcl",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::WebACL",
    },
    "WhiteListIPSet": Object {
      "Properties": Object {
        "Addresses": Array [
          "0.0.0.1/32",
          "0.0.0.2/32",
        ],
        "IPAddressVersion": "IPV4",
        "Name": "WhiteListIPSet",
        "Scope": "CLOUDFRONT",
      },
      "Type": "AWS::WAFv2::IPSet",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
